Tests/MongoSwiftTests/SpecTestRunner/Match.swift

```diff
@@ -0,0 +1,95 @@
+import Foundation
+import MongoSwift
+import Nimble
+
+// swiftlint:disable line_length
+/// Protocol that allows a type to assert it matches a given value according to the specs' MATCHES function.
+/// https://github.com/mongodb/specifications/tree/master/source/connection-monitoring-and-pooling/tests#spec-test-match-function
+internal protocol Matchable {
+    /// Returns whether this MATCHES the expected value according to the function defined in the spec.
+    /// https://github.com/mongodb/specifications/tree/master/source/connection-monitoring-and-pooling/tests#spec-test-match-function
+    func matches(expected: Any) -> Bool
+}
+// swiftlint:enable line_length
+
+/// Extension that adds MATCHES functionality to `Array`.
+extension Array: Matchable {
+    internal func matches(expected: Any) -> Bool {
+        guard !isPlaceholder(expected) else {
```

Comment:
Going by the pseudocode in the spec, if `expected` is 42/"42", we should only verify that `actual` is non-nil, regardless of its type. Only when `expected` isn't 42/"42" do we check the type of `actual` and verify equality/matching and do any special casing.

Reply:
since this check for a placeholder happens at the start of every implementation of `match`, should we consider just checking it in one place e.g. in the nimble matcher?

Reply:
I tend to think of the matchers as wrapper functions rather than things that have implementation specifics within them. (e.g. I wouldn't think putting `self.bsonType == other.bsonType` in `bsonEqual`, even if we had to do it in every implementation of `bsonEquals`).

Instead, I've updated the `Matchable` protocol to require conformers to only implement a `contentMatches` function (which is everything after the placeholder check) and to auto implement `matches` (which does the placeholder check) based on `contentMatches`. This way the placeholder check is done automatically and only in a single place.